File: services/python/quart/src/routes/handlers/internal.py
import structlog
from quart import request
from werkzeug.exceptions import InternalServerError

from routes.errors import AppErrors, send_error


async def internal_error_handler(error: InternalServerError):
    cause = error.original_exception or error

    structlog.get_logger().error(
        "Internal server error",
        method=request.method,
        path=request.path,
        error=repr(cause),
    )

    return send_error(AppErrors.Internal)

File: services/go/chi/routes/middleware/internal.go
package middleware

import (
	"chi/routes"
	"log/slog"
	"net/http"
)

func RecoverMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer recoverPanic(w, r)
		next.ServeHTTP(w, r)
	})
}

func recoverPanic(w http.ResponseWriter, r *http.Request) {
	err := recover()
	if err == nil {
		return
	}
	slog.Error("Internal server error", "method", r.Method, "path", r.URL.Path, "error", err)
	routes.AppErrors.Internal.Error(w)
}
